mbedtls/src/pkcs12/mod.rs

```diff
@@ -0,0 +1,1103 @@
+
+            let hmac_len = Md::hmac(md, &hmac_key, &self.raw_data, &mut computed_hmac)?;
+
+            // FIXME const time compare
+            if computed_hmac[0..hmac_len] != stored_mac[..] {
+                return Err(Pkcs12Error::Custom("Invalid MAC".to_owned()));
+            }
+        }
+
+        return Ok(());
+    }
+
+    /// Decrypt an encrypted Pfx If mac_passphrase is None, it is assumed to be
+    /// identical to the encryption passphrase Dual password PKCS12 files (using
+    /// distinct passwords for encryption and authentication) can be created
+    /// using openssl with the -twopass option.
+    pub fn decrypt(&self, passphrase: &str, mac_passphrase: Option<&str>) -> Pkcs12Result<Pfx> {
+        let passphrase = format_passphrase_for_pkcs12(passphrase);
+
+        if let Some(mac_pass) = mac_passphrase {
+            let mac_passphrase = format_passphrase_for_pkcs12(mac_pass);
+            self.authenticate(&mac_passphrase)?;
+        } else {
+            self.authenticate(&passphrase)?;
+        }
+
+        let new_authsafe = self
+            .authsafe
+            .contents
+            .iter()
+            .map(|data| match data {
+                &ContentInfoContents::Data(ref sc) => {
+                    fn decrypt_pkcs8_sb(sb: &SafeBag, passphrase: &[u8]) -> SafeBag {
+                        if let &SafeBagContents::EncryptedPkcs8(ref p8) = &sb.bag_value {
+                            if let Ok(decrypted_p8) = decrypt_pkcs8(&p8, passphrase) {
+                                return SafeBag {
+                                    bag_id: ObjectIdentifier::from_slice(PKCS12_BAG_KEY),
+                                    bag_value: SafeBagContents::Pkcs8(decrypted_p8),
+                                    bag_attributes: sb.bag_attributes.clone(),
+                                };
+                            }
+                        }
+
+                        return sb.clone();
+                    }
+
+                    let d: Vec<SafeBag> =
+                        sc.0.iter()
+                            .map(|p| decrypt_pkcs8_sb(p, &passphrase))
+                            .collect::<Vec<_>>();
+                    ContentInfoContents::Data(SafeContents(d))
+                }
+                &ContentInfoContents::EncryptedData(ref ed) => {
+                    let decrypted = decrypt_contents(&ed, &passphrase);
+                    if let Ok(sc) = decrypted {
+                        ContentInfoContents::Data(sc)
+                    } else {
+                        ContentInfoContents::EncryptedData(ed.clone())
+                    }
+                }
+            })
+            .collect();
+
+        let decrypted = Pfx {
+            version: self.version,
+            authsafe: ContentInfo {
+                oid: self.authsafe.oid.clone(),
+                contents: new_authsafe,
+            },
+            raw_data: self.raw_data.clone(),
+            macdata: self.macdata.clone(),
```

Comment:
What is the point of cloning these? These fields will never be used again, right?

Reply:
I wanted it to be possible to call decrypt twice without failure. But an easier way to do that is set raw_data to None and have decrypt check for that and return if seen.